### ISSUE TITLE
Python 3.5, building numpy-master with MSVC 14.0 (or visual studio 2015)

### ISSUE THREAD

**Issue by Iceman9** [numpy/numpy#6640]
Hello, I tried building the numpy-master or latest for python 3.5. The reason behind it, because I wish to build it with open-blas, on windows... I know.

Even if I wish to just build it with said MSVC 14.0 I get the following compile error:
numpy\core\src\multiarray\mapping.c(1296): error C2036: 'void *': unknown size

I apologize if I wasted your time over nothing and if the following snippet of code is pasted amateurly. 

To just paste the snippet of the code from numpy\core\src\multiarray\mapping.c , where the problem occurs:
...
npy_intp offset;
...
        _view = (PyArrayObject_)PyArray_NewFromDescr(
                                    Py_TYPE(arr),
                                    fieldtype,
                                    PyArray_NDIM(arr),
                                    PyArray_SHAPE(arr),
                                    PyArray_STRIDES(arr),
                                    PyArray_DATA(arr) + offset,  // <- Here
                                    PyArray_FLAGS(arr),
                                    (PyObject *)arr);
...

From the Data Type API page on docs.scipy.org I read that the npy_intp type is a:
"The enumeration value for a signed integer type which is the same size as a (void *) pointer. This is the type used by all arrays of indices."
I don't have any C/C++ experience and I only found that the MSVC has problems over pointer type void. Also because of that I tried using the casting (char *) as this is supposedly the usual way to sort this out with MSVC. And it didn't work, even though I somehow knew it wouldn't.

Any thoughts on how to deal with this?

**Comment by njsmith** [comment]
Huh, I would have expected that making that '((char*) PyArray_DATA(arr)) +
offset' would have fixed that. Maybe you just need more parentheses?
(Though I think the parentheses shouldn't be necessary because casting
should bind more tightly than addition... it's weird.)
.
In any case thanks for the report, and even more thanks if you manage to
fix it and seen us a pull request with the fix :-). We don't have a good
way to test on windows/msvc at the moment so we mostly rely on folks like
you to do so. Nothing amateurish about your report, though a useful trick
to know about for nicer code pasting on github is to surround the code
block with ``` (three backticks) to prevent github from trying to interpret
the code as if it had formatting characters in it. (This is why github is
trying to italicize part of what you pasted, it has gotten confused by *
characters.) You can google "github flavored markdown" for the full details.

**Comment by Iceman9** [comment]
Hello again! The line you mentioned:  '((char*) PyArray_DATA(arr)) + offset', works, but it isn't in the numpy-master repository. I added this and it compiled until the next problem.

Moving on with the next problem: In the numpy\numpy\core\src\private\mem_overlap.c file the following compile errors arise on line 538,539:

numpy\core\src\private\mem_overlap.c(538): error C2057: expected constant expression
numpy\core\src\private\mem_overlap.c(538): error C2466: cannot allocate an array of constant size 0
numpy\core\src\private\mem_overlap.c(538): error C2133: 'Ep': unknown size
numpy\core\src\private\mem_overlap.c(539): error C2057: expected constant expression
numpy\core\src\private\mem_overlap.c(539): error C2466: cannot allocate an array of constant size 0
numpy\core\src\private\mem_overlap.c(539): error C2133: 'Epsilon': unknown size
numpy\core\src\private\mem_overlap.c(539): error C2133: 'Gamma': unknown size

Basically under unix you have no problem creating arrays in the following term:

```
int some_length = 6;
int array[some_length];
```
While in MSVC that is illegal, since you are passing a non-constant variable as the size parameter for creating an array. So you have to do the old fashioned way with malloc.


The snippet:
UN-ALTERED:
```
diophantine_term_t* Ep[n];
 npy_int64 Epsilon[n], Gamma[n];
```

Altered:
```
diophantine_term_t *Ep = malloc(n * sizeof(diophantine_term_t));
npy_int64 *Epsilon= malloc(n * sizeof(npy_int64));
npy_int64 *Gamma = malloc(n *sizeof(npy_int64));
```

Now with these 2 problems "fixed", the build finishes. Now what follows are tests with the built numpy and then try to build it with openblas, which was built with mingw.

**Comment by njsmith** [comment]
Any chance you could submit a pull request for those changes, so the next
person along didn't have to figure them out from scratch? Also you'll
probably want by to add some calls to free to match the calls to malloc...

**Comment by carlkl** [comment]
the latest Openblas I build with mingwpy is from September:

https://bitbucket.org/carlkl/mingw-w64-for-python/downloads/OpenBLAS_mingwpy_amd64.7z
https://bitbucket.org/carlkl/mingw-w64-for-python/downloads/OpenBLAS_mingwpy-win32.7z
https://bitbucket.org/carlkl/mingw-w64-for-python/downloads/OpenBLAS_mingwpy.txt

If you intend to build a more recent one (OpenBLAS-0.2.15) be sure to link the gcc runtimes statically to avoid the dependency to the gcc runtime DLLs.

**Comment by Iceman9** [comment]
Hey, sorry for not answering in a while, had some work to do. Yeah, I already added the free memory functions, but right now I'm trying to build it with open-blas, well like @carkl said, I have to first statically link the gcc runtimes. I checked the README's, but still I'm doing something wrong.

Also I will publish the pull request I hope soon and hope to not derail this thread off-topic too much.

**Comment by Iceman9** [comment]
So just an update. Building numpy with openblas was "successfull".

 I created the .lib file from the /exports folder in openblas, with lib.exe. Then I put the .dll file from /build/bin and the created .lib file into the openblas folder for numpy. In site.cfg you have to put the full name of "libopenblas", or the link.exe from visual studio won't find them and report the LNK1181 error when trying to link the .lib file.

After applying all the corrections numpy builds. When I installed it to python with 
```python setup.py install```
it reported no errors but when I tried to atlest import numpy I got the following error:
```
Traceback (most recent call last):
  File "<pyshell#0>", line 1, in <module>
    import numpy
  File "C:\Python\lib\site-packages\numpy\__init__.py", line 180, in <module>
    from . import add_newdocs
  File "C:\Python\lib\site-packages\numpy\add_newdocs.py", line 13, in <module>
    from numpy.lib import add_newdoc
  File "C:\Python\lib\site-packages\numpy\lib\__init__.py", line 8, in <module>
    from .type_check import *
  File "C:\Python\lib\site-packages\numpy\lib\type_check.py", line 11, in <module>
    import numpy.core.numeric as _nx
  File "C:\Python\lib\site-packages\numpy\core\__init__.py", line 14, in <module>
    from . import multiarray
ImportError: DLL load failed: The specified module could not be found.
```

**Comment by jaimefrio** [comment]
Those are typically best debugged under Windows using [dependency walker](http://www.dependencywalker.com/), run it on `multiarray.pyd` and you will likely find out what DLL it is having a harrd time finding.

**Comment by Iceman9** [comment]
That's a mighty tool there. As suspected the missing .DDL file is the libopenblas.dll which I forgot to include in numpy main installation directory. Now everything loads fine and in the numpy.__config__.show() shows that I have openblas:
```
>>> np.__config__.show()
blas_mkl_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['libopenblas', 'libopenblas']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    library_dirs = ['g:\\lapack_blas_builds\\numpy-master\\numpy\\numpy\\core\\']
openblas_lapack_info:
    libraries = ['libopenblas', 'libopenblas']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    library_dirs = ['g:\\lapack_blas_builds\\numpy-master\\numpy\\numpy\\core\\']
lapack_opt_info:
    libraries = ['libopenblas', 'libopenblas']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    library_dirs = ['g:\\lapack_blas_builds\\numpy-master\\numpy\\numpy\\core\\']
blas_opt_info:
    libraries = ['libopenblas', 'libopenblas']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    library_dirs = ['g:\\lapack_blas_builds\\numpy-master\\numpy\\numpy\\core\\']
```

It's actually pretty funny, because @carlkl, you were in another thread in which there were the same questions and in the end someone else posted the solution. And because I didn't pay any attention to it, I more or less used so many days wondering what's wrong with the build. I also don't want to start with my try on compiling ATLAS... The thread or discussion: https://groups.google.com/forum/#!topic/openblas-users/EuGTc6bWhZw .  Thanks for the help everyone, I got all the answers and more from you and I will push a request soon.

**Comment by carlkl** [comment]
@Iceman9 , I used libopenblaspy as an alternative name to libopenblas to avoid conflicts with users of Julia, see https://github.com/winpython/winpython/issues/73#issuecomment-94310642
Another question: can you upload your builds to an anaconda.org repository for further tests?

**Comment by Iceman9** [comment]
Question: How would uploading a numpy build with open-blas compiled against my computer benefit the user? If I understand, the easier way would be to simply put the source code up in a way that is compiled and installed with minimum work instead of uploading a .WHL of numpy build I built.

**Comment by mattip** [comment]
@Iceman9 any news with this?

**Comment by carlkl** [comment]
maybe this should be closed as numpy builds with OpenBLAS are available at PYPI.

**Comment by charris** [comment]
Closing. It is old enough that a new issue should be opened if needed.
